HW0/P4/P4.py

```diff
@@ -30,47 +30,106 @@
     # Load true trajectory and plot it
     # Normally, this data wouldn't be available in the real world
     #####################
-
-    # ax.plot(x_coords, y_coords, z_coords,
-    #         '--b', label='True trajectory')
+    s_true = np.loadtxt('P4_trajectory.txt', delimiter=',')
+    ax.plot(s_true[:,0], s_true[:,1], s_true[:,2],
+            '--b', label='True trajectory')
 
     #####################
     # Part 2:
     #
     # Read the observation array and plot it (Part 2)
     #####################
-
-    # ax.plot(x_coords, y_coords, z_coords,
-    #         '.g', label='Observed trajectory')
+    m = np.loadtxt('P4_measurements.txt', delimiter=',')
+    s_obs = m * np.array([1./rx, 1./ry, 1./rz]) 
+    ax.plot(s_obs[:,0], s_obs[:,1], s_obs[:,2],
+            '.g', label='Observed trajectory')
 
     #####################
     # Part 3:
     # Use the initial conditions and propagation matrix for prediction
     #####################
 
-    # A = ?
-    # a = ?
-    # s = ?
+    # Construct matrix A from blocks
+    sub_a = np.eye(3)
+    sub_b = np.eye(3) * dt
+    sub_c = np.zeros((3,3))
+    sub_d = np.eye(3) * (1-c*dt)
+    A = np.bmat([[sub_a, sub_b], [sub_c, sub_d]])
+    A = np.array(A)
+    a = np.array([0.]*5 + [g*dt])
 
     # Initial conditions for s0
     # Compute the rest of sk using Eq (1)
 
-    # ax.plot(x_coords, y_coords, z_coords,
-    #         '-k', label='Blind trajectory')
+    # Initialize s0
+    s0 = s_true[0,:]
+    S_pred0 = np.zeros((6,K))
+    S_pred0[:,0] = s0
+    s = s0
+    for i in range(1,K):
+        s = np.dot(A,s) + a
+        S_pred0[:,i] = s
+
+    ax.plot(S_pred0[0,:], S_pred0[1,:], S_pred0[2,:],
+            '-k', label='Blind trajectory')
 
     #####################
     # Part 4:
     # Use the Kalman filter for prediction
     #####################
 
-    # B = ?
-    # C = ?
+    def predictS(A, s, a):
+        '''Use Eq(2) to compute intermediate s'''
+        return np.dot(A,s) + a
+
+    def predictSig(A, B, Sig):
+        '''
+        Use Eq(3) to compute cov mat.
+        Returns an array.
+        '''
+        # Make sure inputs are matrices
+        A = np.asmatrix(A)
```

Comment:
I would suggest setting up (or converting) all your variables as matrices at the beginning of part 4, running all your calculations on matrices, and then just converting your final result back to an array for graphing purposes.  That way you can avoid having to convert back and forth at each iteration. It probably doesn't make a noticeable difference in run-time for just 120 steps, but it could help speed things up if you were calculating more time steps.